Add price limit for samples

andrew-aisi opened the pull request:
## This PR contains:
- [x] New features
- [ ] Changes to dev-tools e.g. CI config / github tooling
- [ ] Docs
- [ ] Bug fixes
- [ ] Code refactor

### What is the current behavior? (You can also link to an open issue here)

Users cannot set a price based limit for samples (#1152).

### What is the new behavior?

Users can set a price limit for samples if they provide a json file that maps model names to per-token prices for output, cached input, and unique input. As an example, the [model_prices_and_context_window.json](https://github.com/BerriAI/litellm/blob/main/model_prices_and_context_window.json) from litellm provides this information.

### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

I don't believe so. I was worried the new `price` limit type might introduce issues with backwards compatibility in the log viewer but didn't have any issues with the VSCode viewer.

### Other information:

**My goal here is to allow for cost-normalized comparisons, not to ensure end users are actually billed a specified amount.** As such, this design ignores the fact that Inspect could load model responses from its own internal cache. Assuming token prices are correctly specified, a sample should never cost the user more than the limit they've set, but it may cost less due to caching.

I don't like the current setup where the pricing file path is passed all the way to individual samples that each have to read it. I'd prefer to store pricing info globally, but wasn't sure how to best do so - any guidance would be appreciated. 

In this design I was thinking it could be good to support a scenario where multiple models are used for a single sample (e.g., with a multi-agent design), but I'm not sure this is something Inspect actually supports. If not, this could be simplified even further by just calculating the relevant token prices early on and making this available to the solver.

There are a few small things I dislike in the code, will flag via review to get feedback on if/how to improve.

This is related to, but doesn't solve #980. I think it could be extended to do this fairly easily, but wanted to get initial feedback on this approach first. I think that issue might be best solved by tracking actual cost details via API instead of based on tokens, but perhaps 

Leaving this as a draft for now due to the above, but also this still needs:
- [ ] Find a better way to store pricing information globally
- [ ] Documentation
- [ ] Validate web UI changes actually work as intended
- [ ] Think through edge cases in pricing that I've missed and address (e.g., if reasoning token output price differs from regular output tokens)

craigwalton-dsit commented on src/inspect_ai/util/_limit.py:
```diff
@@ -500,6 +556,125 @@ def _validate_message_limit(self, value: int | None) -> None:
+
+        # TODO: validate price format is correct!
+        # TODO: Can we load this file elsewhere instead of in every sample?
+        # TODO: Do we know what the model is at this point? Could multiple models
```
As mentioned in another comment, we do unfortunately have to deal with multiple models. We'll have to store something like a dict (a `defaultdict` comes to mind) keyed by model named to `ModelUsage`, or even just store a running tally of cost (this is ropy if we're using floats though).

Note that you do have access to this from the `record_model_usage()` callsite.

craigwalton-dsit commented on src/inspect_ai/util/_limit.py:
```diff
@@ -500,6 +556,125 @@ def _validate_message_limit(self, value: int | None) -> None:
+
+        super().__init__()
+        if price_file:
+            self._prices = self._load_price_file(price_file)
```
Some would argue we ought to be using `Decimal` for anything finance related to provide exact representations and not suffer from loss of precision. Do you think this would be feasible?

I'm not as worried about things like the limit being defined as a `float` though.

craigwalton-dsit commented on src/inspect_ai/util/_limit.py:
```diff
@@ -500,6 +556,125 @@ def _validate_message_limit(self, value: int | None) -> None:
+            prices = json.load(f)
+
+        # TODO: validate price format is correct!
+        # TODO: Can we load this file elsewhere instead of in every sample?
```
I think this is a good idea, rather than loading + validating the file every time we instantiate a `_PriceLimit`.

I'd lean towards putting this in its own module (e.g. `_cost.py` or `_price.py`).

Something as simple as a global variable in that module would be my starting point. Here's a precedent of this sort of thing https://github.com/UKGovernmentBEIS/inspect_ai/blob/0ae0df980e0866e8c9e9d86c69b84cdb49b35b7f/src/inspect_ai/_util/entrypoints.py#L9

andrew-aisi replied:
I used an lru cache within the _cost.py file to cache reads of the file. But I'm wondering if it might be better to initialize a global with the path to the costs file (and perhaps another global with the contents) instead of passing that path around everywhere.

craigwalton-dsit commented on src/inspect_ai/model/_model.py:
```diff
@@ -1463,6 +1465,7 @@ def record_and_check_model_usage(model: str, usage: ModelUsage) -> None:
     set_model_usage(model, usage, sample_model_usage_context_var.get(None))
     set_model_usage(model, usage, model_usage_context_var.get(None))
     record_model_usage(usage)
+    record_model_usage_price(usage)
```
I think that this `record_and_check_model_usage()` function is only called from `_model.generate()` if the inspect cache was not hit. Which I don't think is what you were hoping for?

https://github.com/UKGovernmentBEIS/inspect_ai/blob/0ae0df980e0866e8c9e9d86c69b84cdb49b35b7f/src/inspect_ai/model/_model.py#L617

I'd recommend double checking this claim though - this is just based on my skimming the source.

craigwalton-dsit commented on src/inspect_ai/util/_limit.py:
```diff
@@ -500,6 +556,125 @@ def _validate_message_limit(self, value: int | None) -> None:
+
+        model_price = self._prices[model.name]
+
+        # TODO: Some models have different pricing for reasoning vs regular
```
Yeah I can imagine this getting pretty complex with the various pricing structures. Definitely worth moving this logic into another module (see other comment).

Note the distinction between inspect caching and provider caching described here https://inspect.aisi.org.uk/caching.html

craigwalton-dsit commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 - Task display: Truncate task config to prevent overflow (collapse dicts, limit individual values to 50 chars, limit overall output to 500 chars).
 - Inspect View: Outline view for transcript which enables high level navigation to solvers, agents, scorers, etc.
 - Inspect View: Fix an issue that prevented the display of the viewer in VSCode when the viewer tab was moved to the background.
+- Price limits: Prevent a sample from exceeding a specified USD-based budget.
```
On the "cost" vs "price" terminology: I prefer "cost" when we're talking about usage or limits. It feels like as you're making calls to models, you're racking up the cost to yourself, not the lab's price (the price remains constant at $/token). I don't object so much to a "price file" - but we might want to also call it "cost file" to have consistent terminology. But I'm totally open to alternative interpretations, but I think we want to nail the naming early on.

> Price is generally used to refer to the amount of money that a seller is asking for something, and cost generally refers to the amount of money that buyer will spend rather than the price that a seller is asking for.
https://www.britannica.com/dictionary/eb/qa/price-and-cost#:~:text=Price%20is%20generally%20used%20to,a%20seller%20is%20asking%20for.

craigwalton-dsit reviewed (commented):
Thanks for pioneering this Andrew, nice job! As requested, I've done a pass over this and have left some initial thoughts as comments.

I think the biggest remaining questions are around token price schemas, validation, whether we support all the different token cost methodologies (e.g. reasoning, cache writes, batched, etc.).

andrew-aisi commented on src/inspect_ai/util/_limit.py:
```diff
@@ -303,6 +305,59 @@ def check_working_limit() -> None:
+    return _PriceLimit(limit, price_file)
+
+
+def record_model_usage_price(usage: ModelUsage) -> None:
```
> I'm also weighing up wether we should merge record_model_usage_price() and record_model_usage().

I've been thinking about this too. I think it could work to just use a single tree tracking token usage and then in the `_CostLimit` and _TokenLimit` implementations check if the limits are exceeded. But I wasn't sure if this would cause problems. Will hack on this and see if I get somewhere.

andrew-aisi commented:
I incorporated the suggested changes and squashed the commits to keep the history from getting too messy. I think this is now working, but it looks like I'll need to fix a few more things to make CI happy.

Beyond the basic functionality, I have a few open questions and could use help writing docs (or I can get to that eventually)
1) Should we show the running/final cost per sample in the UI somewhere? Right now the cost numbers aren't shown or stored anywhere unless the limit is exceeded.
2) The details of some usage (messages and tokens) are stored in the active sample e.g., [here](https://github.com/UKGovernmentBEIS/inspect_ai/pull/1930/commits/75e430f3ad2d15a6aa444d2a2cd93c74a8ab4179#diff-3f032d8d0b05c0185a812aa5be3b6684e906c6f54b6d813748a52eba4d5845edR328) while other usage info (e.g., working time) isn't. It seems somewhat redundant with the main limit code, but I tried to support this in 75e430f3ad2d15a6aa444d2a2cd93c74a8ab4179. Since I'm not sure the point of it, I didn't test it - could remove if this is unnecessary. Maybe it's connected with showing these things in the UI?
3) I'm still a bit confused by caching. I checked the limit was working as expected based on the token usage info I get from `inspect eval` (e.g., with cache reads labeled "CR"). I think this is provider-side caching (which is a discount I do want to account for here), but I didn't get local inspect caching to work. In e8101331d8f87667e93cd369d301634a59610e6f I tried to add support for treating these cache hits as non-free and add a test case, but I got stuck on the test case. Are there any existing tests with multiple models that I could build off?

Thanks and let me know if there's anything else you'd want changed here.

craigwalton-dsit commented on src/inspect_ai/_cli/eval.py:
```diff
@@ -252,6 +253,20 @@ def eval_options(func: Callable[..., Any]) -> Callable[..., click.Context]:
     )
+    @click.option(
+        "--cost-limit",
+        type=float,
```
I wonder if this can be Decimal too? No worries if not.

craigwalton-dsit commented on src/inspect_ai/_view/www/log-schema.json:
```diff
@@ -1395,6 +1395,34 @@
+              "type": "number"
+            },
+            {
+              "type": "string"
```
Is the `string` type required in the schema, or will `number` suffice? Haven't run this myself, so appreciate there may be a good reason for it!

craigwalton-dsit commented on src/inspect_ai/log/_samples.py:
```diff
@@ -149,6 +160,18 @@ def set_active_sample_total_messages(total_messages: int) -> None:
         active.total_messages = total_messages
 
 
+def set_active_sample_cost_limit(cost_limit: Decimal | None) -> None:
```
I think these can be removed if we decide to not show the live cost/cost limit in the TUI (as asked in your question 1.)

craigwalton-dsit commented on src/inspect_ai/model/_model.py:
```diff
@@ -614,6 +617,12 @@ async def generate() -> tuple[ModelOutput, BaseModel]:
+                        total_cost = calculate_model_usage_cost(
+                            {cache_entry.model: existing.usage}
+                        )
+                        record_model_usage_cost(total_cost)
```
Should we also be checking the cost limit here too?

The fact that this takes a different code path to the case where we don't use the local cache highlights that we're taking a different approach with cost to what we do with tokens. Token usage does not include locally cached ones.

Just a note to say I think we'll have to make the distinction crystal clear in the docs or it could lead to surprises.

jjallaire commented:
Cost limit available here: https://inspect.aisi.org.uk/setting-limits.html#cost-limit (thanks for your contribution to this!)
